sound/soc/sof/sof-priv.h

```diff
@@ -118,6 +118,52 @@ struct snd_sof_platform_stream_params {
 	bool no_ipc_position;
 };
 
+struct snd_sof_dsp_ipc_ops {
+	/* doorbell */
+	irqreturn_t (*irq_handler)(int irq, void *context); /* optional */
+	irqreturn_t (*irq_thread)(int irq, void *context); /* optional */
+
+	/* ipc */
+	int (*send_msg)(struct snd_sof_dev *sof_dev,
+			struct snd_sof_ipc_msg *msg); /* mandatory */
+
+	/*
+	 * FW ready checks for ABI compatibility and creates
+	 * memory windows at first boot
+	 */
+	int (*fw_ready)(struct snd_sof_dev *sdev, u32 msg_id); /* mandatory */
+
+	/* host side configuration of the stream's data offset in stream mailbox area */
+	int (*set_stream_data_offset)(struct snd_sof_dev *sdev,
+				      struct snd_pcm_substream *substream,
+				      size_t posn_offset); /* optional */
+
+	/* host read DSP stream data */
+	int (*ipc_msg_data)(struct snd_sof_dev *sdev,
+			    struct snd_pcm_substream *substream,
+			    void *p, size_t sz); /* mandatory */
+
+	void (*ipc_dump)(struct snd_sof_dev *sof_dev); /* optional */
+};
+
+struct snd_sof_dsp_fw_ops {
+	/* FW loading */
+	int (*load_firmware)(struct snd_sof_dev *sof_dev); /* mandatory */
+	int (*load_module)(struct snd_sof_dev *sof_dev,
+			   struct snd_sof_mod_hdr *hdr); /* optional */
+	/* parse platform specific extended manifest, optional */
+	int (*parse_platform_ext_manifest)(struct snd_sof_dev *sof_dev,
+					   const struct sof_ext_man_elem_header *hdr);
+};
+
+struct snd_sof_dsp_trace_ops {
+	int (*trace_init)(struct snd_sof_dev *sdev,
+			  struct sof_ipc_dma_trace_params_ext *dtrace_params); /* optional */
```

Comment:
that's not so good. I don't know how to fix this one.

Reply:
void *params maybe? then we can set struct sof_ipc_dma_trace_params_ext *dtrace_params = params in hda-trace.c hda_dsp_trace_init()

Reply:
The trace_init passes the dtrace_params because
Intel HDA needs to update the stream_tag in the message.
AMD ACP needs to override the buffer.phy_addr (it also provides a stream_tag, but afaik it is not used)
So in essence it is set up the platform side and also to feed back information. How did we do this for the audio streams? They also need the stream_tag, no?

My other take is that the dtrace is SOF (IPC3) specific, IPC4 uses different trace, they also have DMA backed debugging feature, but not sure how that compares to our dtrace...